keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1184 @@
+//
+// Each ClusterTrustBundle object is optionally owned by a particular signer, which
+// updates the trust anchors as the roots of trust for that signer change.
+//
+// A ClusterTrustBundle object wraps a single set of PEM certificates.
+type ClusterTrustBundle struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // The spec of the object.
+    Spec ClusterTrustBundleSpec `json:"spec"`
+}
+
+// ClusterTrustBundleSpec is the spec subfield of a ClusterTrustBundle object.
+type ClusterTrustBundleSpec struct {
+  // The name of the associated signer.
+  //
+  // +optional
+  SignerName string `json:"signerName,omitempty"`
+
+  // The individual trust anchors for this bundle.  A PEM bundle of PEM-wrapped,
+  // DER-formatted X.509 certificate.
+  //
+  // The order of certificates within the bundle has no meaning.
+  PEMTrustAnchors string `json:"pemTrustAnchors"`
+}
+```
+
+#### Access Control
+
+All ClusterTrustBundle objects are assumed to be publicly-readable within the
+cluster, due to the fact that the kubelet trustAnchors volume type will allow
+any pod to mount any ClusterTrustBundle.  This will be backed up by a built-in
+grant of get, list, and watch for ClusterTrustBundle objects to
+`system:authenticated`.
+
+For ClusterTrustBundle objects without `.spec.signerName` specified, no
+nonstandard access control is applied.  Typically, a controller (or human
+operator) who creates such a bundle will grant read access to the
+ClusterTrustBundle to the `system:authenticated` group (tighter grants are
+possible, but are defeated by `pemTrustAnchors` Kubelet volume type, which
+allows any pod in the cluster to read any ClusterTrustBundle).
+
+For ClusterTrustBundle objects with `.spec.signerName` specified, some
+additional admission checks (similar to those on CertificateSigningRequest
+objects) are applied to creates and updates.  Creates and updates are rejected
+unless the requester has permission for the `entrust` verb on group
+`certificates.k8s.io`, resource `signers`, with resourceName equal to
+`.spec.signerName`.
+
+For example, to allow a controller to manage ClusterTrustBundle objects for
+`example.com/my-signer`, the following ClusterRole could be used.
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: example-com-trustmanager
+rules:
+- apiGroups:
+  - certificates.k8s.io
+  resources:
+  - signer
+  resourceNames:
+  - example.com/my-signer
+  verbs:
+  - entrust
+```
+
+(Note that this permission does not need to be granted via RBAC; it could be
+granted via any of Kubernetes' supported authorization methods.)
+
+Updates to ClusterTrustBundle objects that attempt to change `.spec.signerName`
+are always rejected.
+
+#### Well-known ClusterTrustBundles
+
+Kubernetes' built-in signers will export their trust anchors in well-known
+ClusterTrustBundles.
+
+* `kubernetes.io/kube-apiserver-client`: `kubernetes-io--kube-apiserver-client`
```

Comment:
will the kube-apiserver dynamically read and honor these values in its authentication chain?

Reply:
My thinking was that these bundles are read only; the signer implementation is putting them into cluster as a convenience for workloads, and only the signer should be updating them.

So the source of truth in a default Kubernetes install would still be the kube-apiserver and kube-controller-manager flags for these CAs

Reply:
Summarizing a conversation from slack: 

For alpha, I'll just note that these names and prefixes are reserved, and note that for Beta we need to figure out exactly how this will work